### PR TITLE
[SPARK-22737][ML][WIP] OVR transform optimization

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
simplify OVR transform

performance tests:

Data : 153600 instances, 4 features

|numClasses| 3 | 5 | 7 | 10 | 20 | 
|------|----------|------------|----------|------------|----------|
|duration(old) | 6.845616431 | 9.768051372 | 12.500740211 | 17.891663704 | 37.731668899 | 
|duration(new) | 2.502908419 | 2.306560897 | 2.346268255 | 2.411586592 | 2.704394136 | 

## How was this patch tested?
existing tests

### CODE REVIEW

**Comment by zhengruifeng** [comment]
test code:

```
import org.apache.spark.ml.classification._

val df = spark.read.format("libsvm").load("/Users/zrf/Dev/OpenSource/spark/data/mllib/sample_multiclass_classification_data.txt")

val classifier = new LogisticRegression().setMaxIter(1).setTol(1E-6).setFitIntercept(true)
val ovr = new OneVsRest().setClassifier(classifier)

var df2 = df
(0 until 10).foreach{ i => df2 = df2.union(df2) }
df2 = df2.repartition(64)
df2.persist()
df2.count

Seq(3, 5, 7, 10, 20).foreach { k =>
	val dfk = df2.withColumn("label", (rand()*k).cast("int"))
	val ovrModelK = ovr.fit(dfk)
	val start = System.nanoTime
	(0 until 10).foreach { i => ovrModelK.transform(dfk).count }
	val end = System.nanoTime
	val duration = (end - start) / 1e9
	println(s"numClasses $k, duration $duration")
}

```

File: mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala [hunk 1]
```diff
@@ -156,54 +153,22 @@ final class OneVsRestModel private[ml] (
...
-        // add temporary column to store intermediate scores and update
-        val tmpColName = "mbc$tmp" + UUID.randomUUID().toString
-        val updateUDF = udf { (predictions: Map[Int, Double], prediction: Vector) =>
-          predictions + ((index, prediction(1)))
+    val predictUDF = udf { (features: Any) =>
+      var i = 0
+      var maxIndex = Double.NaN
+      var maxPred = Double.MinValue
+      while (i < models.length) {
+        val pred = models(i).predictRawAsFeaturesType(features)(1)
```
Comment (WeichenXu123) [inline mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala]: Here serialize & broadcast models, is it safe and will it bring some overhead ?

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala]: I thought on this, when a model is a distributed model, e.g, this model including a `Dataframe` or `RDD` data, then it cannot be broadcasted. In `OneVsRest`, it should support any kind of classifier, so we cannot suppose that the corresponding model of specified classifier can be broadcast.

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala]: @WeichenXu123  Agree that we should take some distributed classification model like `DistributedLDAModel` in the future into account, but for now all classification models are not distributed.
What about let models tell us whether it is distributed (like adding a method named `isDistributed`), and determine the way of prediction in OVR?

**Comment by zhengruifeng** [comment]
@MLnick  @jkbradley  What's your thoughts?

**Comment by zhengruifeng** [comment]
@mengxr @holdenk How do you think about this? Thanks.

**Comment by zhengruifeng** [comment]
@srowen How do you think about this? Current OVR model's transform is too slow.  Thanks.

**Pull request closed without merging**